roachtest: nil ctx when draining

[(roachtest).sqlsmith/setup=seed-vec/setting=vec failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2550095&tab=buildLog) on [master@6d49a323b52966becfe8a2c38a1a8ccdf8ee58a1](https://github.com/cockroachdb/cockroach/commits/6d49a323b52966becfe8a2c38a1a8ccdf8ee58a1):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed-vec/setting=vec/run_1
	cluster.go:1637,context.go:140,cluster.go:1626,test_runner.go:841: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-2550095-1609311871-12-n4cpu4 --oneshot --ignore-empty-nodes: exit status 1 3: 7027
		4: 7801
		1: 10934
		2: dead
		Error: UNCLASSIFIED_PROBLEM: 2: dead
		(1) UNCLASSIFIED_PROBLEM
		Wraps: (2) attached stack trace
		  -- stack trace:
		  | main.glob..func14
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1147
		  | main.wrap.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:271
		  | github.com/spf13/cobra.(*Command).execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:830
		  | github.com/spf13/cobra.(*Command).ExecuteC
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:914
		  | github.com/spf13/cobra.(*Command).Execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:864
		  | main.main
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1850
		  | runtime.main
		  | 	/usr/local/go/src/runtime/proc.go:204
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
		Wraps: (3) 2: dead
		Error types: (1) errors.Unclassified (2) *withstack.withStack (3) *errutil.leafError

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed-vec/setting=vec](https://teamcity.cockroachdb.com/viewLog.html?buildId=2550095&tab=artifacts#/sqlsmith/setup=seed-vec/setting=vec)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed-vec%2Fsetting%3Dvec.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0xe5c6c2]

goroutine 15240 [running]:
panic(0x41f1260, 0x762c040)
	/usr/local/go/src/runtime/panic.go:1064 +0x545 fp=0xc001fc8ea0 sp=0xc001fc8dd8 pc=0x4e1b65
runtime.panicmem(...)
	/usr/local/go/src/runtime/panic.go:212
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:742 +0x413 fp=0xc001fc8ed0 sp=0xc001fc8ea0 pc=0x4f88b3
github.com/cockroachdb/cockroach/pkg/util/tracing.SpanFromContext(0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/context.go:25 +0x22 fp=0xc001fc8f08 sp=0xc001fc8ed0 pc=0xe5c6c2
github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog(0x0, 0x0, 0x3, 0x0, 0xc00228c6c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:96 +0x39 fp=0xc001fc8f48 sp=0xc001fc8f08 pc=0xf48e79
github.com/cockroachdb/cockroach/pkg/util/log.vEventf(0x0, 0x0, 0xc001fc9000, 0x1, 0x7f4200000002, 0x498e97c, 0x57, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:246 +0x12a fp=0xc001fc9078 sp=0xc001fc8f48 pc=0xf4a3ca
github.com/cockroachdb/cockroach/pkg/util/log.VEvent(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:267
github.com/cockroachdb/cockroach/pkg/sql/execinfra.MisplannedRanges(0x0, 0x0, 0xc0028b8480, 0x1, 0x1, 0x2, 0xc0005815c0, 0x34, 0x34, 0xc001fc9490)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/readerbase.go:65 +0x8c fp=0xc001fc93b0 sp=0xc001fc9078 pc=0x237212c
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*tableReader).generateMeta(0xc000e01000, 0x0, 0x0, 0x7b9088, 0x55613e0, 0xc00290ea80)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/tablereader.go:311 +0x605 fp=0xc001fc9668 sp=0xc001fc93b0 pc=0x25f5f85
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*tableReader).generateTrailingMeta(0xc000e01000, 0x0, 0x0, 0x203000, 0xc000c882a0, 0x54cc960)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/tablereader.go:172 +0x45 fp=0xc001fc96c0 sp=0xc001fc9668 pc=0x25f4c65
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*tableReader).generateTrailingMeta-fm(0x0, 0x0, 0x0, 0x0, 0x9e29cb)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/tablereader.go:171 +0x3e fp=0xc001fc9700 sp=0xc001fc96c0 pc=0x260959e
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).moveToTrailingMeta(0xc000e01000)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:718 +0x226 fp=0xc001fc9840 sp=0xc001fc9700 pc=0x2370ac6
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).MoveToDraining(0xc000e01000, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:611 +0x1b8 fp=0xc001fc9948 sp=0xc001fc9840 pc=0x23703f8
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).ConsumerDone(0xc000e01000)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:905 +0x33 fp=0xc001fc9970 sp=0xc001fc9948 pc=0x2371d73
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).MoveToDraining(0xc0018c4900, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:608 +0x18d fp=0xc001fc9a78 sp=0xc001fc9970 pc=0x23703cd
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).DrainMeta(0xc0018c4900, 0x55613e0, 0xc00290e840, 0x0, 0x0, 0xc002cb84e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:221 +0x4a fp=0xc001fc9b48 sp=0xc001fc9a78 pc=0x27e5caa
github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc.(*Outbox).sendMetadata(0xc0020e85a0, 0x55613e0, 0xc00290e840, 0x7f42b3514b58, 0xc002cb83a0, 0x54c99e0, 0xc002cb8400, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:288 +0x111 fp=0xc001fc9de0 sp=0xc001fc9b48 pc=0x3132711
github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc.(*Outbox).runWithStream(0xc0020e85a0, 0x55613e0, 0xc00290e840, 0x7f42b3514b58, 0xc002cb83a0, 0xc002cb8330)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:328 +0x170 fp=0xc001fc9e58 sp=0xc001fc9de0 pc=0x3132df0
github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc.(*Outbox).Run(0xc0020e85a0, 0x5561320, 0xc002e67a40, 0x54c83a0, 0xc0006d7450, 0x6ebf4d4e00000001, 0xd7741592a040d019, 0xc044e2a3cd, 0x0, 0xc002cb8330, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:180 +0x2f7 fp=0xc001fc9ef8 sp=0xc001fc9e58 pc=0x31321d7
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreator).setupRemoteOutputStream.func1(0x5561320, 0xc002d21b00, 0xc002134600)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:652 +0x125 fp=0xc001fc9f90 sp=0xc001fc9ef8 pc=0x313c065
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1.1(0xc0028b8570, 0x5561320, 0xc002d21b00, 0xc002134600, 0xc002e41638)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1266 +0x44 fp=0xc001fc9fb8 sp=0xc001fc9f90 pc=0x313e984
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc001fc9fc0 sp=0xc001fc9fb8 pc=0x51af41
created by github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1265 +0x6f
```

--- a/pkg/sql/colexec/BUILD.bazel
+++ b/pkg/sql/colexec/BUILD.bazel
@@ -66,6 +66,7 @@ go_library(
         "//pkg/sql/sqlerrors",
         "//pkg/sql/sqltelemetry",  # keep
         "//pkg/sql/types",
+        "//pkg/util",
         "//pkg/util/duration",  # keep
         "//pkg/util/encoding",  # keep
         "//pkg/util/humanizeutil",

--- a/pkg/sql/colexec/colbuilder/execplan.go
+++ b/pkg/sql/colexec/colbuilder/execplan.go
@@ -72,6 +72,10 @@ func wrapRowSources(
 		// necessarily a execinfra.RowSource, so remove the unnecessary
 		// conversion.
 		if c, ok := input.(*colexec.Columnarizer); ok {
+			// Since this Columnarizer has been previously added to Closers and
+			// MetadataSources, this call ensures that all future calls are noops.
+			// Modifying the slices at this stage is difficult.
+			c.MarkAsRemovedFromFlow()
 			toWrapInputs = append(toWrapInputs, c.Input())
 		} else {
 			// Note that this materializer is *not* added to the set of

--- a/pkg/sql/colexec/columnarizer.go
+++ b/pkg/sql/colexec/columnarizer.go
@@ -60,6 +60,11 @@ type Columnarizer struct {
 	accumulatedMeta []execinfrapb.ProducerMetadata
 	ctx             context.Context
 	typs            []*types.T
+
+	// removedFromFlow marks this Columnarizer as having been removed from the
+	// flow. This renders all future calls to Init, Next, Close, and DrainMeta
+	// noops.
+	removedFromFlow bool
 }
 
 var _ colexecop.Operator = &Columnarizer{}
@@ -136,6 +141,9 @@ func newColumnarizer(
 
 // Init is part of the Operator interface.
 func (c *Columnarizer) Init() {
+	if c.removedFromFlow {
+		return
+	}
 	// We don't want to call Start on the input to columnarizer and allocating
 	// internal objects several times if Init method is called more than once, so
 	// we have this check in place.
@@ -149,6 +157,9 @@ func (c *Columnarizer) Init() {
 
 // Next is part of the Operator interface.
 func (c *Columnarizer) Next(context.Context) coldata.Batch {
+	if c.removedFromFlow {
+		return coldata.ZeroBatch
+	}
 	var reallocated bool
 	switch c.mode {
 	case columnarizerBufferingMode:
@@ -232,6 +243,9 @@ var (
 
 // DrainMeta is part of the MetadataSource interface.
 func (c *Columnarizer) DrainMeta(ctx context.Context) []execinfrapb.ProducerMetadata {
+	if c.removedFromFlow {
+		return nil
+	}
 	c.MoveToDraining(nil /* err */)
 	for {
 		meta := c.DrainHelper()
@@ -245,6 +259,9 @@ func (c *Columnarizer) DrainMeta(ctx context.Context) []execinfrapb.ProducerMeta
 
 // Close is part of the Operator interface.
 func (c *Columnarizer) Close(ctx context.Context) error {
+	if c.removedFromFlow {
+		return nil
+	}
 	c.InternalClose()
 	return nil
 }
@@ -273,4 +290,14 @@ func (c *Columnarizer) Child(nth int, verbose bool) execinfra.OpNode {
 // Input returns the input of this columnarizer.
 func (c *Columnarizer) Input() execinfra.RowSource {
 	return c.input
+}
+
+// MarkAsRemovedFromFlow is called by planning code to make all future calls on
+// this columnarizer noops. It exists to support an execution optimization where
+// a Columnarizer is removed from a flow in cases where it would be the input to
+// a Materializer (which is redundant). Simply bypassing the Columnarizer is not
+// enough because it is added to a slice of Closers and MetadataSources that are
+// difficult to change once physical planning moves on from the Columnarizer.
+func (c *Columnarizer) MarkAsRemovedFromFlow() {
+	c.removedFromFlow = true
 }

--- a/pkg/sql/colexec/columnarizer.go
+++ b/pkg/sql/colexec/columnarizer.go
@@ -21,6 +21,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
 	"github.com/cockroachdb/cockroach/pkg/sql/rowenc"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/errors"
 )
 
@@ -117,7 +118,15 @@ func newColumnarizer(
 		processorID,
 		nil, /* output */
 		nil, /* memMonitor */
-		execinfra.ProcStateOpts{InputsToDrain: []execinfra.RowSource{input}},
+		execinfra.ProcStateOpts{
+			InputsToDrain: []execinfra.RowSource{input},
+			TrailingMetaCallback: func(ctx context.Context) []execinfrapb.ProducerMetadata {
+				if err := c.Close(ctx); util.CrdbTestBuild && err != nil {
+					// Close never returns an error.
+					colexecerror.InternalError(errors.AssertionFailedf("unexpected error %v from Columnarizer.Close", err))
+				}
+				return nil
+			}},
 	); err != nil {
 		return nil, err
 	}
@@ -132,6 +141,7 @@ func (c *Columnarizer) Init() {
 	// we have this check in place.
 	if c.initStatus == colexecop.OperatorNotInitialized {
 		c.accumulatedMeta = make([]execinfrapb.ProducerMetadata, 0, 1)
+		c.ctx = c.StartInternalNoSpan(c.ctx)
 		c.input.Start(c.ctx)
 		c.initStatus = colexecop.OperatorInitialized
 	}
@@ -235,7 +245,7 @@ func (c *Columnarizer) DrainMeta(ctx context.Context) []execinfrapb.ProducerMeta
 
 // Close is part of the Operator interface.
 func (c *Columnarizer) Close(ctx context.Context) error {
-	c.input.ConsumerClosed()
+	c.InternalClose()
 	return nil
 }
 

--- a/pkg/sql/colexec/columnarizer_test.go
+++ b/pkg/sql/colexec/columnarizer_test.go
@@ -80,34 +80,53 @@ func TestColumnarizerDrainsAndClosesInput(t *testing.T) {
 	evalCtx := tree.MakeTestingEvalContext(st)
 	defer evalCtx.Stop(ctx)
 
-	rb := distsqlutils.NewRowBuffer([]*types.T{types.Int}, nil /* rows */, distsqlutils.RowBufferArgs{})
 	flowCtx := &execinfra.FlowCtx{
 		Cfg:     &execinfra.ServerConfig{Settings: st},
 		EvalCtx: &evalCtx,
 	}
 
-	const errMsg = "artificial error"
-	rb.Push(nil, &execinfrapb.ProducerMetadata{Err: errors.New(errMsg)})
-	c, err := NewBufferingColumnarizer(ctx, testAllocator, flowCtx, 0 /* processorID */, rb)
-	require.NoError(t, err)
+	for _, tc := range []struct {
+		name           string
+		consumerClosed bool
+	}{
+		{
+			name:           "ConsumerClosed",
+			consumerClosed: true,
+		}, {
+			name:           "ConsumerDone",
+			consumerClosed: false,
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			const errMsg = "artificial error"
+			rb := distsqlutils.NewRowBuffer([]*types.T{types.Int}, nil /* rows */, distsqlutils.RowBufferArgs{})
+			rb.Push(nil, &execinfrapb.ProducerMetadata{Err: errors.New(errMsg)})
+			c, err := NewBufferingColumnarizer(ctx, testAllocator, flowCtx, 0 /* processorID */, rb)
+			require.NoError(t, err)
 
-	c.Init()
+			c.Init()
 
-	// If the metadata is obtained through this Next call, the Columnarizer still
-	// returns it in DrainMeta.
-	err = colexecerror.CatchVectorizedRuntimeError(func() { c.Next(ctx) })
-	require.True(t, testutils.IsError(err, errMsg), "unexpected error %v", err)
+			// If the metadata is obtained through this Next call, the Columnarizer still
+			// returns it in DrainMeta.
+			err = colexecerror.CatchVectorizedRuntimeError(func() { c.Next(ctx) })
+			require.True(t, testutils.IsError(err, errMsg), "unexpected error %v", err)
 
-	// Calling DrainMeta from the vectorized execution engine should propagate to
-	// non-vectorized components as calling ConsumerDone and then draining their
-	// metadata.
-	meta := c.DrainMeta(ctx)
-	require.True(t, len(meta) == 0)
-	require.True(t, rb.Done)
-	require.Equal(t, execinfra.DrainRequested, rb.ConsumerStatus, "unexpected consumer status %d", rb.ConsumerStatus)
-	// Closing the Columnarizer should call ConsumerClosed on the processor.
-	require.NoError(t, c.Close(ctx))
-	require.Equal(t, execinfra.ConsumerClosed, rb.ConsumerStatus, "unexpected consumer status %d", rb.ConsumerStatus)
+			if tc.consumerClosed {
+				// Closing the Columnarizer should call ConsumerClosed on the processor.
+				require.NoError(t, c.Close(ctx))
+				require.Equal(t, execinfra.ConsumerClosed, rb.ConsumerStatus, "unexpected consumer status %d", rb.ConsumerStatus)
+			} else {
+				// Calling DrainMeta from the vectorized execution engine should propagate to
+				// non-vectorized components as calling ConsumerDone and then draining their
+				// metadata.
+				meta := c.DrainMeta(ctx)
+				require.True(t, len(meta) == 0)
+				require.True(t, rb.Done)
+				require.Equal(t, execinfra.DrainRequested, rb.ConsumerStatus, "unexpected consumer status %d", rb.ConsumerStatus)
+			}
+			require.True(t, c.Closed, "the ProcessorBase.Closed bool should be set in either case")
+		})
+	}
 }
 
 func BenchmarkColumnarize(b *testing.B) {

--- a/pkg/sql/execinfra/processorsbase.go
+++ b/pkg/sql/execinfra/processorsbase.go
@@ -711,6 +711,15 @@ func (pb *ProcessorBase) moveToTrailingMeta() {
 			pb.trailingMeta = append(pb.trailingMeta, execinfrapb.ProducerMetadata{TraceData: trace})
 		}
 	}
+
+	if util.CrdbTestBuild && pb.Ctx == nil {
+		panic(
+			errors.AssertionFailedf(
+				"unexpected nil ProcessorBase.Ctx when draining. Was StartInternal called?",
+			),
+		)
+	}
+
 	// trailingMetaCallback is called after reading the tracing data because it
 	// generally calls InternalClose, indirectly, which switches the context and
 	// the span.
@@ -864,11 +873,30 @@ func ProcessorSpan(ctx context.Context, name string) (context.Context, *tracing.
 //   < other initialization >
 // so that the caller doesn't mistakenly use old ctx object.
 func (pb *ProcessorBase) StartInternal(ctx context.Context, name string) context.Context {
+	return pb.startImpl(ctx, true /* createSpan */, name)
+}
+
+// StartInternalNoSpan does the same as StartInternal except that it does not
+// start a span. This is used by pass-through components whose goal is to be a
+// silent translation layer for components that actually do work (e.g. a
+// planNodeToRowSource wrapping an insertNode, or a columnarizer wrapping a
+// rowexec flow).
+func (pb *ProcessorBase) StartInternalNoSpan(ctx context.Context) context.Context {
+	return pb.startImpl(ctx, false /* createSpan */, "")
+}
+
+func (pb *ProcessorBase) startImpl(
+	ctx context.Context, createSpan bool, spanName string,
+) context.Context {
 	pb.origCtx = ctx
-	pb.Ctx, pb.span = ProcessorSpan(ctx, name)
-	if pb.span != nil && pb.span.IsVerbose() {
-		pb.span.SetTag(execinfrapb.FlowIDTagKey, pb.FlowCtx.ID.String())
-		pb.span.SetTag(execinfrapb.ProcessorIDTagKey, pb.processorID)
+	if createSpan {
+		pb.Ctx, pb.span = ProcessorSpan(ctx, spanName)
+		if pb.span != nil && pb.span.IsVerbose() {
+			pb.span.SetTag(execinfrapb.FlowIDTagKey, pb.FlowCtx.ID.String())
+			pb.span.SetTag(execinfrapb.ProcessorIDTagKey, pb.processorID)
+		}
+	} else {
+		pb.Ctx = ctx
 	}
 	pb.EvalCtx.Context = pb.Ctx
 	return pb.Ctx

--- a/pkg/sql/plan_node_to_row_source.go
+++ b/pkg/sql/plan_node_to_row_source.go
@@ -116,9 +116,7 @@ func (p *planNodeToRowSource) SetInput(ctx context.Context, input execinfra.RowS
 }
 
 func (p *planNodeToRowSource) Start(ctx context.Context) {
-	// We do not call p.StartInternal to avoid creating a span. Only the context
-	// needs to be set.
-	p.Ctx = ctx
+	ctx = p.StartInternalNoSpan(ctx)
 	p.params.ctx = ctx
 	if !p.started {
 		p.started = true

